src/emulation_core/mips/datapath.rs

```diff
@@ -230,37 +237,135 @@ impl MipsDatapath {
 
     /// Decode an instruction into its individual fields.
     fn instruction_decode(&mut self) {
-        // TODO: Use an enum and structs rather than individual fields.
-        self.opcode = (self.instruction >> 26) & 0b111111;
-        self.rs = (self.instruction >> 21) & 0b11111;
-        self.rt = (self.instruction >> 16) & 0b11111;
-        self.rd = (self.instruction >> 11) & 0b11111;
-        self.shamt = (self.instruction >> 6) & 0b11111;
-        self.funct = self.instruction & 0b111111;
-        self.imm = self.instruction & 0xFFFF;
+        let op: u32 = self.instruction >> 26;
+        match op {
+            0 => {
+                self.instruction_enum = Instruction::RType(RType {
+                    op: ((self.instruction >> 26) & 0x3F) as u8,
+                    rs: ((self.instruction >> 21) & 0x1F) as u8,
+                    rt: ((self.instruction >> 16) & 0x1F) as u8,
+                    rd: ((self.instruction >> 11) & 0x1F) as u8,
+                    shamt: ((self.instruction >> 6) & 0x1F) as u8,
+                    funct: (self.instruction & 0x3F) as u8,
+                })
+            }
+
+            // ORI, feels kinda scummy to put this here
```

Comment:
Might not be all bad here. We could change the binary `0b001101` to a defined constant, then have many values on that one match case for all the I-type instructions.

Like `OPCODE_ORI | OPCODE_ANOTHER | OPCODE_THIRD => { ...`